### PR TITLE
Drop support for ruby 2.1

### PR DESCRIPTION
**Pull request by jaredbeck**
Ruby 2.1 reached EoL on 2017-04-01.
https://www.ruby-lang.org/en/news/2017/04/01/support-of-ruby-2-1-has-ended/

PT continued to support it for 6 months after EoL.

### CODE REVIEW

**Comment by jaredbeck** [comment]
In https://github.com/airblade/paper_trail/pull/995 Ben asked:

> Is there something that Ruby 2.1 breaks, just out of curiosity?

Good question, and I'm glad you're asking it :) The way I phrase the question is, what burden does supporting an EoL ruby put on the library? Then, we can weigh that against the needs of people using EoL rubies.

First, supporting ruby 2.1 makes testing slightly more difficult because rails no longer supports ruby 2.1, as of rails 5. TravisCI makes this easy though, so that's only a slight burden.

Second, I'd like to eventually use the new features of ruby, like `Hash#dig`, frozen string literals, and the safe navigation operator. (Those are all from 2.3. Ruby 2.2 and 2.4 are mostly performance improvements). Supporting EoL rubies prevents us from using these new features.

People using EoL rubies can always continue to use PT <= 7, and we can continue to backport bug fixes to the 7 series. I'm happy to do that.

Basically, I'm happy to support EoL rails and ruby versions for a little while (6 months to a year) but I'd argue for a policy where we eventually drop support.

**Comment by batter** [comment]
I was just saying, that I don't know that we necessarily need to officially support it, however, leaving the `required_ruby_version` minimum threshold at `2.1` likely doesn't hurt, and I'm guessing people could still continue using the modern versions without much problem.  That being said, I know that some of the modern versions of `ActiveRecord` / rails, which this gem depends on, also require more modern versions of Ruby, so perhaps we should match what they do in terms of required ruby version?

**Comment by jaredbeck** [comment]
> .. leaving the required_ruby_version minimum threshold at 2.1 likely doesn't hurt ..

I agree that we could probably get away with `required_ruby_version("2.1")` for a while, even if we're not testing against 2.1. I'd prefer not to, because I would be concerned about claiming a level of compatibility that we don't test for, but if you want to it's OK with me.

In the future, when we begin using ruby 2.3 syntax (e.g. safe nav. operator) we will have to specify `required_ruby_version("2.3")`.

> That being said, I know that some of the modern versions of ActiveRecord / rails, which this gem depends on, also require more modern versions of Ruby, so perhaps we should match what they do in terms of required ruby version?

Makes sense to me. :+1: PT is a rails plugin, so it makes sense to track the same versions of ruby that rails depends on, within reason. One exception: rails 4.2, which is still living, supports ruby 1.9.3 and we already dropped support for 1.9.3. But otherwise, I think that policy makes sense going forward.

**Comment by batter** [comment]
> I agree that we could probably get away with required_ruby_version("2.1") for a while, even if we're not testing against 2.1. I'd prefer not to, because I would be concerned about claiming a level of compatibility that we don't test for, but if you want to it's OK with me.

You know what, I actually don't really care that much either way. Historically my opinion would be to say leave it unless we know something's broken, but on the other hand if someone wants to use an older ruby version they can always fork the gem and change it themselves and the specification will at least have warned them there could be an issue.

> In the future, when we begin using ruby 2.3 syntax (e.g. safe nav. operator) we will have to specify required_ruby_version("2.3").

Agreed, and that is a nice operator that they've added 😄 

👍  to everything else

**Comment by jaredbeck** [comment]
> .. if someone wants to use an older ruby version they can always fork the gem and change it themselves ...

Or just continue to use PT 7.

> 👍 to everything else

Cool, thanks.
